src/include/duckdb/common/file_system.hpp

```diff
@@ -130,9 +132,121 @@ class FileSystem {
+	                                FileLockType lock = FileLockType::NO_LOCK) override {
+		return FindFileSystem(path)->OpenFile(path, flags, lock);
+	}
+
+	virtual void Read(FileHandle &handle, void *buffer, int64_t nr_bytes, idx_t location) override {
+		handle.file_system.Read(handle, buffer, nr_bytes, location);
+	};
+
+	virtual void Write(FileHandle &handle, void *buffer, int64_t nr_bytes, idx_t location) override {
+		handle.file_system.Write(handle, buffer, nr_bytes, location);
+	}
+
+	int64_t Read(FileHandle &handle, void *buffer, int64_t nr_bytes) override {
+		return handle.file_system.Read(handle, buffer, nr_bytes);
+	}
+
+	int64_t Write(FileHandle &handle, void *buffer, int64_t nr_bytes) override {
+		return handle.file_system.Write(handle, buffer, nr_bytes);
+	}
+
+	int64_t GetFileSize(FileHandle &handle) override {
+		return handle.file_system.GetFileSize(handle);
+	}
+	time_t GetLastModifiedTime(FileHandle &handle) override {
+		return handle.file_system.GetLastModifiedTime(handle);
+	}
+
+	void Truncate(FileHandle &handle, int64_t new_size) override {
+		handle.file_system.Truncate(handle, new_size);
+	}
+
+	void FileSync(FileHandle &handle) override {
+		handle.file_system.FileSync(handle);
+	}
+
+	// need to look up correct fs for this
+	bool DirectoryExists(const string &directory) override {
+		return FindFileSystem(directory)->DirectoryExists(directory);
+	}
+	void CreateDirectory(const string &directory) override {
+		FindFileSystem(directory)->CreateDirectory(directory);
+	}
+
+	void RemoveDirectory(const string &directory) override {
+		FindFileSystem(directory)->RemoveDirectory(directory);
+	}
+
+	bool ListFiles(const string &directory, std::function<void(string, bool)> callback) override {
+		return FindFileSystem(directory)->ListFiles(directory, callback);
+	}
+
+	void MoveFile(const string &source, const string &target) override {
+		FindFileSystem(source)->MoveFile(source, target);
+	}
+
+	bool FileExists(const string &filename) override {
+		return FindFileSystem(filename)->FileExists(filename);
+	}
+
+	virtual void RemoveFile(const string &filename) override {
+		FindFileSystem(filename)->RemoveFile(filename);
+	}
+
+	vector<string> Glob(string path) override {
+		return FindFileSystem(path)->Glob(path);
+	}
+
+	// these goes to the default fs
+	void SetWorkingDirectory(string path) override {
+		default_fs.SetWorkingDirectory(path);
+	}
+
+	string GetWorkingDirectory() override {
+		return default_fs.GetWorkingDirectory();
+	}
+
+	string GetHomeDirectory() override {
+		return default_fs.GetWorkingDirectory();
+	}
+
+	idx_t GetAvailableMemory() override {
+		return default_fs.GetAvailableMemory();
+	}
+
+	void RegisterProtocolHandler(string protocol, unique_ptr<FileSystem> protocol_fs) override {
+		protocol_handler_fss[protocol] = move(protocol_fs);
+	}
+
+private:
+	FileSystem *FindFileSystem(string path) {
+		for (auto &handler : protocol_handler_fss) {
+			if (path.rfind(handler.first, 0) == 0) {
```

Comment:
`memcmp` should work here as well no? Or am I missing something?

Reply:
This did not seem like something worth optimising given the low number of entries in map. `rfind` seems to be implemented fairly efficiently: https://github.com/llvm/llvm-project/blob/3696227c10f5e5841223c2a2fb63fdd1d50a7930/libcxx/include/__string#L1016

Reply:
Ah fair enough, indeed that seems like it essentially does a memcmp under the hood when `pos=0`.